### PR TITLE
managed privnet: the code

### PR DESCRIPTION
**Pull request by greut**
cf. #183 

- [x] `exoscale_network`
- [x] `exoscale_nic`
- [x] `exoscale_network_test`
- [x] `exoscale_nic_test`
- [x] the errors messages returned by the API breaks the library (fix to be done over there)

### CODE REVIEW

**Comment by greut** [comment]
Testers wanted:
https://sos-ch-dk-2.exo.io/yoanyoan/terraform-provider-exoscale_v0.9.36.1_x-551f477.bz2

**Comment by greut** [comment]
Another one:
https://sos-ch-dk-2.exo.io/yoanyoan/terraform-provider-exoscale_v0.9.36.2_x-b9d3557.bz2

**Comment by greut** [comment]
- create it (hopefully it works the first time)
- change the static IPs (as below)
- run it many times and watch it never converge

```diff
   # static IP address
-  ip_address = "${format("10.0.0.%d", count.index + 1)}"
+  ip_address = "${format("10.0.0.%d", count.index + 3)}"

```

**Comment by greut** [comment]
```
=== RUN   TestProvider
--- PASS: TestProvider (0.00s)
=== RUN   TestProvider_impl
--- PASS: TestProvider_impl (0.00s)
=== RUN   TestAccAffinityGroup
--- PASS: TestAccAffinityGroup (5.17s)
=== RUN   TestAccCompute
--- PASS: TestAccCompute (340.22s)
=== RUN   TestAccDomainRecord
--- PASS: TestAccDomainRecord (7.49s)
=== RUN   TestAccDomain
--- PASS: TestAccDomain (2.99s)
=== RUN   TestAccElasticIP
--- PASS: TestAccElasticIP (7.77s)
=== RUN   TestAccNetwork
--- PASS: TestAccNetwork (9.25s)
=== RUN   TestAccNic
--- PASS: TestAccNic (58.01s)
=== RUN   TestAccSecondaryIP
--- PASS: TestAccSecondaryIP (173.47s)
=== RUN   TestAccSecurityGroupRule
--- PASS: TestAccSecurityGroupRule (5.34s)
=== RUN   TestAccSecurityGroup
--- PASS: TestAccSecurityGroup (0.87s)
=== RUN   TestAccSSHKeyPair
--- PASS: TestAccSSHKeyPair (0.56s)
=== RUN   TestValidateIPv4StringNumber
--- PASS: TestValidateIPv4StringNumber (0.00s)
=== RUN   TestValidateIPv4StringNonIP
--- PASS: TestValidateIPv4StringNonIP (0.00s)
=== RUN   TestValidateIPv4StringOk
--- PASS: TestValidateIPv4StringOk (0.00s)
=== RUN   TestValidateIPv4StringKo
--- PASS: TestValidateIPv4StringKo (0.00s)
=== RUN   TestValidateIPv6StringNumber
--- PASS: TestValidateIPv6StringNumber (0.00s)
=== RUN   TestValidateIPv6StringNonIP
--- PASS: TestValidateIPv6StringNonIP (0.00s)
=== RUN   TestValidateIPv6StringKo
--- PASS: TestValidateIPv6StringKo (0.00s)
=== RUN   TestValidateIPv6StringOk
--- PASS: TestValidateIPv6StringOk (0.00s)
PASS
coverage: 61.2% of statements
ok      github.com/exoscale/terraform-provider-exoscale/exoscale        611.188s        coverage: 61.2% of statements 
```

**Comment by greut** [comment]
https://sos-ch-dk-2.exo.io/yoanyoan/terraform-provider-exoscale_SNAPSHOT-9fac51c_linux_amd64.tar.gz

File: examples/managed-private-network/README.md [hunk 1]
```diff
@@ -0,0 +1,59 @@
...
+
+```
+$ terraform apply
+```
+
+Be patient and wait for the instances to be ready.
+
+## Inspecting the private network network
+
+Connecting to any of them and use `nmap`, to inspect what in the private
```
Comment (mcorbin) [inline examples/managed-private-network/README.md]: typo: "what is in"

File: exoscale/resource_network.go [hunk 2]
```diff
@@ -317,36 +282,16 @@ func applyNetwork(d *schema.ResourceData, network *egoscale.Network) error {
...
-	} else {
-		d.Set("dns1", "")
-	}
-
-	if network.DNS2 != nil {
-		d.Set("dns2", network.DNS2.String())
-	} else {
-		d.Set("dns2", "")
+		d.Set("start_ip", nil)
+		d.Set("end_ip", nil)
```
Comment (mcorbin) [inline exoscale/resource_network.go]: "netmask" should not be set to nil ?

Comment (greut) [reply exoscale/resource_network.go]: right, I couldn't do it because of the defaultValue but that didn't work. :+1:

**Comment by greut** [comment]
https://sos-ch-dk-2.exo.io/yoanyoan/terraform-provider-exoscale_SNAPSHOT-8ee95cb_linux_amd64.tar.gz

**Comment by greut** [comment]
@pierre-emmanuelJ @mcorbin a release candidate.

https://sos-ch-dk-2.exo.io/yoanyoan/terraform-provider-exoscale_SNAPSHOT-db487b7_darwin_amd64.tar.gz
https://sos-ch-dk-2.exo.io/yoanyoan/terraform-provider-exoscale_SNAPSHOT-db487b7_linux_amd64.tar.gz

**Comment by greut** [comment]
> ok      github.com/exoscale/terraform-provider-exoscale/exoscale        534.480s       coverage: 61.5% of statements

**Comment by mcorbin** [comment]
I made more tests with the provider. 

When a nic has a static ip, If I remove the static IP I have:

```
Terraform will perform the following actions:

  ~ exoscale_nic.eth_static
      ip_address: "10.0.0.14" => ""


Plan: 0 to add, 1 to change, 0 to destroy.

Do you want to perform these actions?
  Terraform will perform the actions described above.
  Only 'yes' will be accepted to approve.

  Enter a value: yes

exoscale_nic.eth_static: Modifying... (ID: d5265d28-027d-489e-9089-2a2bd41b53a9)
  ip_address: "10.0.0.14" => ""

Error: Error applying plan:

1 error(s) occurred:

* exoscale_nic.eth_static: 1 error(s) occurred:

* exoscale_nic.eth_static: UpdateVMNicIP.ipaddress (slice) is required, got zero IPv4 address

Terraform does not automatically rollback in the face of errors.
Instead, your Terraform state file has been partially updated with
any resources that successfully completed. Please address the error
above and apply again to incrementally change your infrastructure.
```
In that case, I think the nic should me recreated.

**Comment by greut** [comment]
> In that case, I think the nic should me recreated.

why?

**Comment by greut** [comment]
@mcorbin new behaviour is the following.

> [ERROR] new value of "ip_address" cannot be empty. old value was 10.0.0.1. The resource must be recreated instead.

**Comment by mcorbin** [comment]
Ok, thanks ;)

**Comment by greut** [comment]
thanks for your time.

**Pull request merged**